Persons, events, announcements (#4101)
PostHog/posthog.com: contents/blog/persons-on-events.md

---
date: 2022-08-29
title: 'How we’re improving performance by combining persons and events’'
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

In [our last product update](/blog/the-posthog-array-1-39-0) we announced a beta for a substantial change to the way we handle persons and events on PostHog. Today, after gathering your feedback and seeing the improvements to query performance, we’re rolling this change out to all PostHog users. 

This change essentially combines [persons](/manual/persons) and [events](/manual/events) into a single ClickHouse table, moving person IDs and properties _onto_ events. You won’t see any UI changes — persons will still have their own Persons & Groups section on the sidebar, for example — and there’ll be no loss of functionality, but here are some changes you may see:

> - **Faster results for queries involving person properties and events**. Putting persons and events into a single ClickHouse table means we no longer have to join tables to get results on queries involving these data. As a result, query performance will improve by up to 400%.

Author: @macobo Do we have any real figures we could use here?

> - **Faster results for queries involving person properties and events**. Putting persons and events into a single ClickHouse table means we no longer have to join tables to get results on queries involving these data. As a result, query performance will improve by up to 400%.

Reviewer: It's called person-on-events, but actually applies to groups as well, so it's not just person properties, but group aggregation & group properties as well

> - **Faster filtering of events with person properties**. Similarly, filtering events by person properties is much faster when a join is no longer necessary. Anywhere where you’re working with persons and events, PostHog will be faster. 

Reviewer: My tweak seems to suggest it's only true when all 3 are present, which isn't the case. Hmmm.

```suggestion
- **Faster filtering of events with person properties**. Similarly, filtering events by person properties is much faster when a join is no longer necessary. Anywhere where you’re working with persons, groups, and events, PostHog will be faster. 
```

> - **Anonymous and identified users will no longer be merged**. Previously, when a user was identified through an event, such as logging in to a service, we would end up with separate records for unidentified and identified behavior. We would then [merge these into a single person](//docs/how-posthog-works/ingestion-pipeline#merging-two-persons). This could sometimes create conflicts and [other performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Avoiding this merge means some users may be double-counted, but is much more performant and mostly [mitigated by the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer).

Reviewer: Heading seems incorrect.

This is true only in cases where the identified user becomes anonymous _again_.

cc: Ingestion team for better wording here, it shouldn't go in as-is.

Author: Thanks Neil! I've tweaked the language here, but will wait for a review from @EDsCODE or @tiina303 before progressing. :)

> - **You can create queries based on person properties at the time of an event.** This wasn’t previously possible and was often requested. Now, with this change, 

Reviewer: incomplete?

Also worth noting: If you want queries to be based on present person properties, you can do so via creating a cohort with the relevant person properties & filtering on that cohort.

> - **Self-hosting users will see a storage increase**. This is due to extra information being stored in your self-hosted ClickHouse. We expect users to see their storage requirement increase by approximately XX. 

Author: Would also like to get a figure here.

> Self-hosting users also need to manually upgrade deployment to take advantage of this change, while the change will be automatically targeted for users on PostHog Cloud. If you’re self-hosting PostHog, you need to run [async migration](/docs/runbook/async-migrations) number 0006. Afterwards, change your `PERSON_ON_EVENTS_ENABLED` [instance setting](/docs/self-host/configure/instance-settings) to `true`. You'll need to be [a Staff user](/docs/self-host/configure/instance-settings#staff-users) to make changes.

Reviewer: The instance setting change depends on when this is going out. It will become automatic in 1.41 I think, so we want to message something like:

"You must run this migration before version 1.41". (cc: ingestion team for the accurate wording / plan for full release!)

Reviewer 2: > to manually upgrade deployment

this makes it sound different from normal operational flow, I'd just say "need to update to the latest version of PostHog"

For the other part maybe sth like: This will be enabled with version 1.41, however if you'd like to check it out earlier you can do so by enabling the <flag info> after running the 0006 async migration.

Author: As we'll be holding this back for the full release, I think we can just cut the second part entirely. Thanks!

---
date: 2022-08-29
title: 'How we’re improving performance by combining persons and events’'
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

> In [a previous product update](/blog/the-posthog-array-1-39-0) we announced a beta for a substantial change to the way we handle persons and events on PostHog. Today, after gathering your feedback and seeing the improvements to query performance, we’re rolling this change out to all PostHog users. 

Author: ```suggestion
In [a previous product update](/blog/the-posthog-array-1-39-0) we announced a beta for a substantial change to the way we handle persons and events on PostHog. Today, after gathering your feedback and seeing the improvements to query performance, we’re rolling this change out to all PostHog Cloud users, and making it available for Self-Hosted users. 
```

> In [a previous product update](/blog/the-posthog-array-1-39-0) we announced a beta for a substantial change to the way we handle persons and events on PostHog. Today, after gathering your feedback and seeing the improvements to query performance, we’re rolling this change out to all PostHog users. 

Reviewer 2: fyi it's default on after next release for self-hosted (after they run 0007)

This change essentially combines [persons](/manual/persons) and [events](/manual/events) into a single ClickHouse table, adding person IDs and properties _onto_ events. This change also applies to groups, adding group properties and aggregation to the same table. 

You won’t see any UI changes — persons will still have their own Persons & Groups section on the sidebar, for example — and there’ll be no loss of functionality, but here are some changes you may see:

- **Faster results for queries involving person or group properties and events**. Putting persons, groups and events into a single ClickHouse table means we no longer have to join tables to get results on queries involving these data. As a result, query performance will improve by up to 400%.

- **Faster filtering of events with person or group properties**. Similarly, filtering events by person or group properties is much faster when a join is no longer necessary. Anywhere where you’re working with persons or groups, and events, PostHog will be faster. 

> - **Users will no longer be merged retroactively in some situations**. Previously, when an identified user logged in from a different browser (thus becoming anonymous) we would end up with separate records for their unidentified and identified behavior. We would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons), which could create conflicts and [performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Our new approach avoids this merge and is much more performant. However, this means the anonymous events are attributed to a different, now non-existent person.

Reviewer 2: This feels a bit off to me. We always do the merge for the person (but don't change events), the difference is previously at query time we'd do joins, so the events would be tied to the same person. New approach we look at events at query time and they would for the anonymous events be tied to a non-existent person, but we'd show it as a separate unique person in the trend for example.

```suggestion
- **Users will no longer be merged retroactively in some situations**. When an identified user logged in from a different browser (thus becoming anonymous) we end up with separate records for their unidentified and identified behavior. Upon identifying we would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons). Previously, since at query time we would join with person info, all these events from before and after would be tied to the same person. Our new approach, where we write person info into events, (and since events are not changeable) the anonymous events are attributed to a different, now non-existent person, see [consequences-of-merging](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging).
```

Author: ```suggestion
- **Users will no longer be merged retroactively in some situations**. When an identified user logs in from a different browser (thus becoming anonymous) we end up with separate records for their unidentified and identified behavior, which are then merged into a single person. We used to do this using database joins at query time, so that all events were tied to the same person. Now, we simply look at the events when queries are created - meaning that anonymous events are shown as separate, unique persons in insights.
```
@tiina303 Is this correct? As clear as I could make it. 

This is the one point in all this which I'm a little worried about, however. If we're essentially double-counting a bunch of users then I can see that causing our users to have less faith in our data...

Reviewer 2: This is still a bit off specifically `which are then merged into a single person. We used to do this using database joins at query time, so that all events were tied to the same person.` 
 We always do the merge - so future events will be tied to the same person (but we don't edit past events and never have). The change is to queries however - previously we joined at query time. Now we just use the values in the events table (which if we merged have a two person_ids before the merge that would show up as 2 unique persons in trends)

also maybe `which are then merged into a single person via $identify or $create_alias` - as merging only happens if someone uses one of those two.

- **You can create insights based on person properties at the time of an event.** This wasn’t previously feasible and was often requested. Now, with this change, it's possible!

> - **Self-hosting users will see a storage increase**. This is due to extra information being stored in your self-hosted ClickHouse. We expect users to see their storage requirement increase by approximately XX. 

Author: @tiina303 @macobo Is there any way to get an estimate figure for this point?

Reviewer 2: it depends on how many props the use on persons & groups, that varies a lot, so not really, best we could do would be to describe how to look up the data sizes average for their persons and provide some estimate from that, but it's complex and I don't think anyone would use it and it's hard to get this to be accurate in any way too.

Author: ```suggestion
- **Self-hosting users will see a storage increase**. This is due to extra information being stored in your self-hosted ClickHouse. It's difficult to give an estimate on this as the impact will vary a lot depending on configuration and usage per organization. 
```

> Self-hosting users will need to [update to the latest version of PostHog](/docs/runbook/upgrading-posthog), while the change will be automatically targeted for users on PostHog Cloud. 

Author: ```suggestion
Self-hosting users will need to [update to the latest version of PostHog](/docs/runbook/upgrading-posthog), while the change will be automatically rolled out for users on PostHog Cloud. 
```

While improving performance has been a major motivation for this change, this is also a crucial step in ensuring PostHog can continue to scale effectively. 

> Since launching in 2020, PostHog has been adopted by over 15,000 companies and has tracked over 50 billion events. In order to ensure that current and future users have the best possible experience, we need our systems to work as efficiently as possible. Moving persons on to events is a key to this, along with supporting work such as [using materialized columns in ClickHouse](/blog/clickhouse-materialized-columns) to speed up queries even further.

Author: ```suggestion
Since launching in 2020, PostHog has been adopted by over 15,000 companies and has tracked over 50 billion events. In order to ensure that current and future users have the best possible experience, we need our systems to work as efficiently as possible. Moving persons on to events is an important part of this, along with supporting work such as [using materialized columns in ClickHouse](/blog/clickhouse-materialized-columns) to speed up queries even further.
```

> Since launching in 2020, PostHog has been adopted by over 15,000 companies and has tracked over 50 billion events. In order to ensure that current and future users have the best possible experience, we need our systems to work as efficiently as possible. Moving persons on to events is an important part of this, along with supporting work such as [using materialized columns in ClickHouse](/blog/clickhouse-materialized-columns) to speed up queries even further.

Reviewer 2: ```suggestion
Since launching in 2020, PostHog has been adopted by over 15,000 companies and has tracked over 50 billion events. In order to ensure that current and future users have the best possible experience, we need our systems to work as efficiently as possible. Adding persons on to events is an important part of this, along with supporting work such as [using materialized columns in ClickHouse](/blog/clickhouse-materialized-columns) to speed up queries even further.
```

> This said, even though the change is now being fully deployed, we’re still eager to hear [your feedback](https://www.g2.com/g2gives/girls-who-code-pillar-2022/and/posthog) and understand how we can keep making PostHog better. If you’d like more information about any of the changes above then we’re happy to [answer your questions](/questions).

Reviewer 2: Not sure if we want to link to feedback, I worry that we'll see issues and complaints as we've already seen problems in the past. Also the feedback link girls-who-code

Author: ```suggestion
This said, even though the change is now being fully deployed, we’re still eager to hear [your feedback](/slack) and understand how we can keep making PostHog better. If you’d like more information about any of the changes above then we’re happy to [answer your questions](/questions).
```

---
date: 2022-08-29
title: 'How we’re improving performance by combining persons and events’'
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
categories: ['Release notes', 'Product updates']
featuredImage: ../images/blog/posthog-array-blog.png
featuredImageType: full
---

In [our last product update](/blog/the-posthog-array-1-39-0) we announced a beta for a substantial change to the way we handle persons and events on PostHog. Today, after gathering your feedback and seeing the improvements to query performance, we’re rolling this change out to all PostHog users. 

> This change essentially combines [persons](/manual/persons) and [events](/manual/events) into a single ClickHouse table, moving person IDs and properties _onto_ events. This change also applies to groups, adding group properties and aggregation to the same table. 

Reviewer 2: We're still keeping the person properties separately too, e.g. when you go to the person page you'd see that
```suggestion
This change essentially combines [persons](/manual/persons) and [events](/manual/events) into a single ClickHouse table, adding person IDs and properties _onto_ events. This change also applies to groups, adding group properties and aggregation to the same table. 
```

You won’t see any UI changes — persons will still have their own Persons & Groups section on the sidebar, for example — and there’ll be no loss of functionality, but here are some changes you may see:

- **Faster results for queries involving person or group properties and events**. Putting persons, groups and events into a single ClickHouse table means we no longer have to join tables to get results on queries involving these data. As a result, query performance will improve by up to 400%.

- **Faster filtering of events with person or group properties**. Similarly, filtering events by person or group properties is much faster when a join is no longer necessary. Anywhere where you’re working with persons or groups, and events, PostHog will be faster. 

> - **Users will no longer be merged in some situations**. Previously, when an anonymous user was identified through an event, such as logging in to a service, and then becomes anonymous _again_ we would end up with separate records for unidentified and identified behavior. We would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons), which could create conflicts and [performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Our new approach avoids this merge and is much more performant. We mitigate the risk of double-counting such users with [the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer)

Author: ```suggestion
- - **Users will no longer be merged retroactively in some situations**. Previously, when an identified user logged in from a different browser (thus becoming anonymous) we would end up with separate records for unidentified and identified behavior. We would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons), which could create conflicts and [performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Our new approach avoids this merge and is much more performant. However, this means the anonymous events are attributed to a different, now non-existent person. We mitigate the risk of double-counting such users with [the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer).
```

> - **Users will no longer be merged in some situations**. Previously, when an anonymous user was identified through an event, such as logging in to a service, and then becomes anonymous _again_ we would end up with separate records for unidentified and identified behavior. We would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons), which could create conflicts and [performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Our new approach avoids this merge and is much more performant. We mitigate the risk of double-counting such users with [the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer)

Reviewer: Magical, thank you!

> - - **Users will no longer be merged retroactively in some situations**. Previously, when an identified user logged in from a different browser (thus becoming anonymous) we would end up with separate records for unidentified and identified behavior. We would then [merge these into a single person](/docs/how-posthog-works/ingestion-pipeline#merging-two-persons), which could create conflicts and [performance issues](/docs/how-posthog-works/ingestion-pipeline#consequences-of-merging). Our new approach avoids this merge and is much more performant. However, this means the anonymous events are attributed to a different, now non-existent person. We mitigate the risk of double-counting such users with [the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer).

Reviewer 2: > We mitigate the risk of double-counting such users with [the event buffer](/docs/how-posthog-works/ingestion-pipeline/#1-event-buffer).

This is not accurate - the event buffer only helps us merge initial login, if a backend event happened to come in within 60s & that's it's only goal, any later logins (for events before the login/identify) we won't do any merging in the future

Author: Cutting the last sentence to clarify this. LMK if there's anything further to add here?

> - **You can create insights based on person properties at the time of an event.** This wasn’t previously feasible and was often requested. Now, with this change, it's possible!

Reviewer 2: note that this is also the default now, not sure if we want to call it out or not
